### PR TITLE
Implement feedback from integrating a new `nucypher-core` version into `nucypher`

### PR DESCRIPTION
**Pull request by piotr-roslaniec**
Based on #119. Please review it first.

**Type of PR:**
- Feature

**Required reviews:** 
- 1

**What this does:**
- Fixes naming conflict with `umbral.PublicKey`
- Exposes a utility for making `ferveo` Python module, to be used in `nucypher-core`

**Why it's needed:**
- Need to take these changes into account before proceeding with https://github.com/nucypher/nucypher/pull/3135 and before releasing the stable `ferveo` version

**Notes for reviewers:**
- Please consider reviewing https://github.com/nucypher/nucypher/pull/3135 first

### CODE REVIEW

File: ferveo/Cargo.toml [hunk 1]
```diff
@@ -47,7 +47,8 @@ console_error_panic_hook = { version = "0.1.7", optional = true }
 getrandom = { version = "0.2", features = ["js"], optional = true }
 js-sys = { version = "0.3.63", optional = true }
 wasm-bindgen = { version = "0.2.86", optional = true }
-wasm-bindgen-derive = { version = "0.2.0", optional = true }
+# TODO: Replace with a new release after this issue is fixed: https://github.com/fjarri/wasm-bindgen-derive/issues/3
+wasm-bindgen-derive = { git = "https://github.com/piotr-roslaniec/wasm-bindgen-derive.git", rev = "d13475a9b36d6c6e5f2f99372d175f21b00bf0d6", optional = true }
```
Comment (derekpierre) [inline ferveo/Cargo.toml:51]: I guess we couldn't just use an older version of `wasm-bindgen-derive` here for now ... ?

Comment (piotr-roslaniec) [reply ferveo/Cargo.toml:51]: I don't know if that would help. I think the root cause of the issue in question is the incompatibility of `wasm-bindgen` used through our dependencies with the new Rust stable release, 1.70.

Comment (derekpierre) [reply ferveo/Cargo.toml:51]: So would things potentially work with rust versions < 1.70 or perhaps some other sub-dependency causes issues anyway? I'm unfamiliar with the specifics of Rust [crate] release patterns. Would that imply that we need to modify the MSRV to be 1.70?

Basically, I'm wondering if supporting 1.70 is something that is absolutely necessary (based on Rust best-practicies) or can be done down the road with us only supporting older versions for now. It was only released yesterday.

Kind of like in Python how we only support 3.8, 3.9, 3.10 but not 3.11 at the moment. Apologies if I'm totally missing the boat here.

Comment (piotr-roslaniec) [reply ferveo/Cargo.toml:51]: I think you have the right idea, let me just flesh things out a bit:

> So would things potentially work with rust versions < 1.70

Yes, but then we would have to drop the support for `stable` because technically we wouldn't support the `stable` release. 

> some other sub-dependency causes issues anyway

We could spend a bit more time diagnosing the issue (https://github.com/fjarri/wasm-bindgen-derive/issues/3), but my priority was to introduce a workaround for breaking change in our build. I'm putting a high premium on the ability to ship code, even if it means adding technical debt in the form of patches to the project. This is why I opted to fork & fix.

> Basically, I'm wondering if supporting 1.70 is something that is absolutely necessary 

It's strongly preferred to keep up with the `stable` release. Rust has a policy of no-breaking changes and relies on editions instead (see below). So the crate ecosystem should work in a way where if I release a crate with Rust `n` version anyone should be able to use it with `n-1` version of Rust* - Because _assume_ there won't be any breaking changes.

MSRV is here so that the inter-dependant libraries can figure out if they can build on each other.

* Assuming my code doesn't do [something funny](https://utcc.utoronto.ca/~cks/space/blog/programming/Rust1BackwardIncompatibility)

So in short if something doesn't work on `stable` then it's highly probable there is an unintended breaking change/bug somewhere.

>  Kind of like in Python how we only support 3.8, 3.9, 3.10 but not 3.11 at the moment. 

Python versions are more like [Rust editions](https://doc.rust-lang.org/edition-guide/editions/), i.e. Rust 2018, Rust 2021.

Comment (derekpierre) [reply ferveo/Cargo.toml]: 👍 - Thanks for the helpful background info.

**Review by derekpierre — APPROVED** [review]
🎸
